fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1685,15 +1715,16 @@ private void saveStoreHeader(@Nonnull RecordMetaDataProto.DataStoreInfo storeHea
         if (recordStoreState == null) {
             throw new RecordCoreException("cannot update store header with a null record store state");
         }
-        recordStoreState.beginWrite();
+        beginRecordStoreStateWrite();
         try {
             context.setDirtyStoreState(true);
-            synchronized (recordStoreState) {
-                recordStoreState.setStoreHeader(storeHeader);
-                ensureContextActive().set(getSubspace().pack(STORE_INFO_KEY), storeHeader.toByteArray());
-            }
+            recordStoreState.updateAndGet(state -> {
+                state.setStoreHeader(storeHeader);
+                return state;
+            });
+            ensureContextActive().set(getSubspace().pack(STORE_INFO_KEY), storeHeader.toByteArray());
```

Comment:
I think these two statements still need to be in a synchronized block (on the `recordStoreStateRef` rather than the `recordStoreState`, actually). In particular, if two people are updating the store header at once, without the synchronized block, one thread could (1) update the in-memory state's store header with the new value then (2) the other thread could update the value in memory and then update the value in the database and then (3) the first thread could wake back up and write the value to the database (so the in-memory answer to what the store header is would be different to the in-database answer).

Reply:
There's another synchronized block like this on 2.8 that will need to get similarly updated. That's just a note for later, at the moment.